core/include/clist.h

```diff
@@ -98,23 +171,175 @@ static inline clist_node_t *clist_remove_head(clist_node_t *list)
+ *
+ * @param[in,out]   list        Pointer to the *list* to remove last element
+ *                              from.
+ */
+static inline clist_node_t *clist_rpop(clist_node_t *list)
+{
+    if (list->next) {
+        list_node_t *last = list->next;
+        while (list->next->next != last) {
+            clist_lpoprpush(list);
+        }
+        return clist_lpop(list);
+    }
+    else {
+        return NULL;
+    }
+}
+
+/**
+ * @brief Finds node and returns its predecessor
+ *
+ * @note Complexity: O(n)
+ *
+ * @param[in]       list    pointer to clist
+ * @param[in,out]   node    Node to look for
+ *                          Must not be NULL.
+ *
+ * @returns         predecessor of node if found
+ * @returns         NULL if node is not a list member
+ */
+static inline clist_node_t *clist_find_before(const clist_node_t *list, const clist_node_t *node)
+{
+    clist_node_t *pos = list->next;
+    if (!pos) {
+        return NULL;
+    }
+    do {
+        pos = pos->next;
+        if (pos->next == node) {
+            return pos;
+        }
+    } while (pos != list->next);
+
+    return NULL;
+}
+
+/**
+ * @brief Finds and returns node
+ *
+ * @note Complexity: O(n)
+ *
+ * @param[in]       list    pointer to clist
+ * @param[in,out]   node    Node to look for
+ *                          Must not be NULL.
+ *
+ * @returns         node if found
+ * @returns         NULL if node is not a list member
+ */
+static inline clist_node_t *clist_find(const clist_node_t *list, const clist_node_t *node)
+{
+    clist_node_t *tmp = clist_find_before(list, node);
+    if (tmp) {
+        return tmp->next;
+    }
+    else {
+        return NULL;
+    }
+}
+
+/**
+ * @brief Finds and removes node
+ *
+ * @note Complexity: O(n)
+ *
+ * @param[in]       list    pointer to clist
+ * @param[in,out]   node    Node to remove for
+ *                          Must not be NULL.
+ *
+ * @returns         node if found and removed
+ * @returns         NULL if node is not a list member
+ */
+static inline clist_node_t *clist_remove(clist_node_t *list, clist_node_t *node)
+{
+    if (list->next) {
+        if (list->next->next == node) {
+            return clist_lpop(list);
+        }
+        else {
+            clist_node_t *tmp = clist_find_before(list, node);
+            if (tmp) {
+                tmp->next = tmp->next->next;
+                if (node == list->next) {
+                    list->next = tmp;
+                }
+                return node;
+            }
+        }
+    }
+
+    return NULL;
+}
+
+/**
+ * @brief Traverse clist, call function for each member
+ *
+ * If @p func returns non-zero, traversal will be aborted like when calling
+ * break within a for loop.
+ *
+ * @param[in]       list        List to traverse.
+ * @param[in]       func        Function to call for each member.
+ */
+static inline void clist_foreach(clist_node_t *list, int(*func)(clist_node_t *))
```

Comment:
I wrote in an earlier comment that this function should be inlined to allow the compiler to optimize for the case where `func` is compile time constant, which is likely the most common case.
(the comment was deleted by github when the file no longer exists)